components/script/dom/visibilitystateentry.rs

```diff
@@ -0,0 +1,71 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::ops::Deref;
+
+use dom_struct::dom_struct;
+
+use crate::dom::bindings::codegen::Bindings::DocumentBinding::DocumentVisibilityState;
+use crate::dom::bindings::codegen::Bindings::PerformanceEntryBinding::PerformanceEntryMethods;
+use crate::dom::bindings::codegen::Bindings::VisibilityStateEntryBinding::VisibilityStateEntryMethods;
+use crate::dom::bindings::num::Finite;
+use crate::dom::bindings::reflector::reflect_dom_object;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::performanceentry::PerformanceEntry;
+
+#[dom_struct]
+pub struct VisibilityStateEntry {
+    entry: PerformanceEntry,
+}
+
+impl VisibilityStateEntry {
+    fn new_inherited(state: DocumentVisibilityState, timestamp: f64) -> VisibilityStateEntry {
+        let name = match state {
+            DocumentVisibilityState::Visible => DOMString::from("visible"),
+            DocumentVisibilityState::Hidden => DOMString::from("hidden"),
+        };
+        VisibilityStateEntry {
+            entry: PerformanceEntry::new_inherited(
+                name,
+                DOMString::from("visibility-state"),
+                timestamp,
+                0.,
+            ),
+        }
+    }
+
+    #[allow(crown::unrooted_must_root)]
```

Comment:
I think you can move this to `new_inherited`  if you create the entry within the call to `reflect_`